ports/rp2/machine_uart.c

```diff
@@ -272,7 +410,8 @@ STATIC mp_uint_t machine_uart_write(mp_obj_t self_in, const void *buf_in, mp_uin
             }
             MICROPY_EVENT_POLL_HOOK
         }
-        uart_get_hw(self->uart)->dr = *src++;
+        ringbuf_put(&(self->write_buffer), *src++);
```

Comment:
I think there's a (rare) race condition here:
- ringbuf is full so while loop above is waiting for an empty slot
- some long-running callback is executed during `MICROPY_EVENT_POLL_HOOK`
- UART TX IRQ drains the ring buffer
- UART TX IRQ has nothing left to send so no more TX IRQs are generated
- while loop above finishes
- ringbuf_put on this line adds more to the ringbuf
- data does not get put out on UART because there are no more TX IRQs

Reply:
That's indeed a clever observation. I can add another call to 'uart_fill_tx_fifo(self, true)`, which will start sending the remaining characters. If the FIFO is already full, it will do nothing.

Reply:
I think the issue is still there.  The `uart_fill_tx_fifo()` call needs to be within this loop, otherwise `while (ringbuf_free(...)) { ... }` will time out if it hits this scenario.

For efficiency, probably best to put `uart_fill_tx_fifo()` within that wait-ringbuf-free loop.

Reply:
Right. I faintly recall having had it there once.

Reply:
Updated. Meanwhile I have a working (and failing) test case showing the need for the write lock in addition to the test for the read lock.

Reply:
Ok, that's great you can test these subtle conditions.

Does it make sense to put these tests in to this repo, somehow?  If they are generic `machine.UART` tests then they could be used to test any port's UART.

At least, I would like to run the tests, if it's possible to share what you did?

Reply:
Sure. It is a little script with both a recv and send call. I use two RP2's for that test. Script below. It's a little bit hacky. It sends blocks of '0123456789', and the receiver tests for consistency. You run recv() on one board and send() on the other. As soon as you disable the read lock, you will get recv errors. It takes a while. They are not very frequent. When you disable the write lock, you will also get receive errors, but that more frequent.
```
from machine import UART
import time

def send(baud=460800):

    u = UART(0, baud, tx=Pin(12), rx=Pin(13), txbuf=1024, timeout_char = 1000, timeout=1000)
    # u = UART(1, baud, tx=Pin(8), rx=Pin(9), txbuf=200, timeout_char = 1000, timeout=1000)
    m="0123456789"
    sum = 0
    count=0
    try:
        while True:
            for i in range(5):
                sum += u.write(m*20)
                time.sleep_ms(2)
            print(sum)
            time.sleep(1)
            count += 1
    except KeyboardInterrupt:
        pass

def recv(baud=460800):
    u = UART(0, baud, tx=Pin(12), rx=Pin(13), rxbuf=1024)
    # u = UART(1, baud, tx=Pin(8), rx=Pin(9), rxbuf=1024)

    u.read(u.any())
    sum = 0
    try:
        nullerror = 0
        seqerror = 0
        enderror = 0
        while True:
            new = u.any()
            if new:
                while new:
                    sum += new
                    msg = u.read(new)
                    print(msg)
                    start = msg[0]
                    for c in msg[1:]:
                        if ((start + 1) != c) and not (start == 57 and c == 48):
                            if c != 0:
                                seqerror += 1
                            else:
                                nullerror += 1
                        start = c
                    time.sleep_ms(1)
                    new = u.any()
                if msg[-1] != 57:  # must be 9
                    enderror += 1
                print(msg, "\n", sum, seqerror, nullerror, enderror)

    except KeyboardInterrupt:
        if u.any():
            print(u.read(u.any()))
    print(sum)
```

Reply:
Thanks!  I ran this code for about 13Mbytes of transfer and saw no problems.